presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -598,17 +612,50 @@ else if (redistributeWrites) {
                     !source.getProperties().isCompatibleTablePartitioningWith(shufflePartitioningScheme.get().getPartitioning(), false, metadata, session) &&
                     !(source.getProperties().isRefinedPartitioningOver(shufflePartitioningScheme.get().getPartitioning(), false, metadata, session) &&
                             canPushdownPartialMerge(source.getNode(), partialMergePushdownStrategy))) {
+                PartitioningScheme exchangePartitioningScheme = shufflePartitioningScheme.get();
+                if (node.getTablePartitioningScheme().isPresent() && isPrestoSparkAssignBucketToPartitionForPartitionedTableWriteEnabled(session)) {
+                    int writerThreadsPerNode = getTaskPartitionedWriterCount(session);
+                    int bucketCount = getBucketCount(node.getTablePartitioningScheme().get().getPartitioning().getHandle());
+                    int[] bucketToPartition = new int[bucketCount];
```

Comment:
is `bucketToPartition.length` always equal to bucket count  ?

Reply:
Yes. Basically the length of the `bucketToPartition` is equal to the total number of buckets. Then each slot in this array represents a single bucket, and the value assigned set's the partition id where this bucket has to get processed.